kernel-test-nohz: Add postinst/postrm script fragments for configuring CPU isolation

gratian opened the pull request:
The NO_HZ_FULL kernel ptest relies on CPU isolation for running an
undisturbed RT polling test.

Add a postinst script fragment that changes the kernel boot parameters via
the 'othbootargs' variable to include the necessary 'isolcpus' and
'nohz_full' parameters configured to use the last CPU available on the
system under test. These parameters isolate that CPU and flag it for
stopping the scheduler tick whenever possible.

The corresponding postrm script fragment undoes changes to 'othbootargs'
(which is normally empty).

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian committed:
kernel-test-nohz: Add postinst/postrm script fragments for configuring CPU isolation

The NO_HZ_FULL kernel ptest relies on CPU isolation for running an
undisturbed RT polling test.

Add a postinst script fragment that changes the kernel boot parameters via
the 'othbootargs' variable to include the necessary 'isolcpus' and
'nohz_full' parameters configured to use the last CPU available on the
system under test. These parameters isolate that CPU and flag it for
stopping the scheduler tick whenever possible.

The corresponding postrm script fragment undoes changes to 'othbootargs'.

Signed-off-by: Gratian Crisan <gratian.crisan@ni.com>

gratian commented:
FYI: @ni/rtos

billpittman commented on recipes-kernel/kernel-tests/kernel-test-nohz.bb:
```diff
@@ -31,4 +31,20 @@ do_install_ptest_append() {
+}
+
+pkg_postrm_${PN}-ptest_append() {
+    fw_setenv othbootargs
```
I'm concerned that this would unset other othbootargs from other ipks as well as this one.

gratian replied:
That's fair. In practice I don't expect this test to ever get uninstalled (rather the test system gets re-formatted) but let me see if I can do better.

billpittman replied:
Ah right, I forgot this was a ptest.  Then in that case, do better if you want, or simply leave it.  In fact you could even remove the postrm all together.  There's no real harm in leaving othbootargs "dirty" in a test where the system gets formatted.

gratian replied:
I've added some sed magic in v2 to only add/remove the 'isolcpus' and 'nohz_full' parameters.

amstewart commented on recipes-kernel/kernel-tests/kernel-test-nohz.bb:
```diff
@@ -31,4 +31,20 @@ do_install_ptest_append() {
+    ISOLATED_CPU=$((CPUS - 1))
+
+    if [ $ISOLATED_CPU -gt 0 ]; then
+        fw_setenv othbootargs 'isolcpus='$ISOLATED_CPU' nohz_full='$ISOLATED_CPU
```
Does this imply that the ptest requires a reboot between package installation and executing the test?

In that case, I think it would be more appropriate to include a tool with the ptest subpackage which the user can intentionally run and insert these options into their cmdline (or run-in-reverse to remove them). Setting them at package-install time seems inappropriate to me.

gratian replied:
Yes, this test requires a reboot; however:

- changing the kernel command line is not the only reason a reboot is required for this test. The main reason is it depends on and installs a new kernel (configured for NO_HZ_FULL). Basically even w/o these postinst/postrm scripts a reboot will be required.
- the ATS this is going to be run in already has a reboot in place after the install.

I would prefer to keep this as a postinst script because it greatly simplifies the main use case: running this ptest as part of the ATS. With the current implementation all the ATS needs to do is install this ptest, reboot and run ptest-runner. If I start to introduce custom scripts that means the ptest is no longer self contained and the ATS framework now needs to know that for this particular ptest additional scripts need to be called (to change the command line). This adds unneeded complication.

For somebody running this test by hand I can print a message advising about a reboot being required. I expect manually running this test to be a corner case that's not encountered very often outside the ATS.

gratian commented:
Changes since V1:

-  I've added some sed magic to preserve the contents of 'othbootargs' on postinst/postrm and only append or remove the 'isolcpus' and 'nohz_full' parameters.

- I've tested that this functions correctly and doesn't clobber any parameters that might have been added by the user before or after installing this test.
